### PR TITLE
[v2.7] Add Configuration Hooks for ProvV2

### PR DESCRIPTION
**Pull request by HarrisonWAffel**
## Problem
When provisioning nodes using the RHEL operating system the node will fail to fully provision due to two NetworkManager services interfering with the CNI's configuration. This results in the node indefinitely staying in the `Waiting for probes` stage (or some other similar failure condition). This issue was fixed for RKE1 clusters via change in `rancher-machine` (see https://github.com/rancher/rancher/issues/36526). However, this solution is not applicable for ProvV2 as all configuration is done via cloud-init. 
 
## Solution
Allow the user to specify predefined configuration _hooks_ within the yaml of a machine pool. These hooks denote particular commands that will be added to the install script passed to the downstream node. Currently, cloud-init is used to install the rancher system-agent for ProvV2. This PR adds additional commands after the installation of the agent is complete. 

The contents of each hook are defined within Rancher source code, ensuring that arbitrary commands cannot be injected within the cloud-init script. The included hook performs the same configuration step that is performed when provisioning RKE1 clusters using RHEL, which allows the default RHEL AMI to be used without any prior configuration. More hooks can be added as needed, so other pre-configuration issues with particular OS's can be automatically resolved. 

When building the install script, Rancher does not know what particular OS is being used, so these hooks must be manually configured. This differs from the behavior of RKE1 provisioning, however from what I am able to tell it cannot be avoided. 

To use these new hooks the user must modify the YAML of a particular machine pool to denote which hooks should be applied to the install script. For example, the following YAML will disable the `nm-cloud-setup.service` and `nm-cloud-setup.timer` using `systemctl`.

```yaml
machinePools:
    - controlPlaneRole: true
      etcdRole: true
      hooks:
      - disable-network-manager
      machineConfigRef:
        kind: <CONFIG_TYPE>
        name: <CONFIG_NAME>
      name: <POOL_NAME>
      quantity: 1
      unhealthyNodeTimeout: 0s
      workerRole: true
```
 
Specifying this hook will add the following logic to the install script, after the system-agent is installed.
```
sudo systemctl list-units --all | grep -Fq nm-cloud-setup.timer; if [ $? -eq 0 ]; then sudo systemctl disable nm-cloud-setup.timer; fi
sudo systemctl list-units --all | grep -Fq nm-cloud-setup.service; if [ $? -eq 0 ]; then sudo systemctl disable nm-cloud-setup.service; fi
```

Machines must be reprovisioned/recreated if this hook is added to the YAML after a machine has already been created in an infrastructure provider. 

## Testing
+ Begin to create an RKE2 cluster within the Rancher UI
+ Create three machine pools
   + Two of the pools will use a RHEL AMI. One pool should have the `etcd` and `control-plane` roles and the other should just have the `worker` role. 
   + One of the pools will use the default Ubuntu AMI. This pool should just have the `worker` role. 
+ After creating the three pools, click on the `Edit as YAML` button in the lower right hand corner
+ Add the `disable-network-manager` hook to the first pool configured with a RHEL AMI, do not define any hooks for the other two pools
+ Create the cluster after modifying the YAML 
+ Observe that the RHEL pool with the hook defined provisions and becomes available, and that the RHEL pool which does not have the hook defined does not come up or become available. Ensure that the Ubuntu pool comes up and becomes without the need to specify any hooks. 

### Manual Testing
I've done the above testing and have used the new hook to provision a new RHEL 8.6 cluster. I observed it come up and become useable. Machine pools within the same cluster which did not have this hook specified did not come up successfully, as expected. Machine pools which used OS's which did not need any pre-configuration were provisioned and became available as expected.

## QA Testing Considerations
This fix should address many RHEL versions which do not provision successfully by default, so testing should not be limited to just RHEL 8.6. 
 
### Regressions Considerations
None that I can think of.

### CODE REVIEW

**Review by jakefhyde — CHANGES_REQUESTED** [review]

**Review by jakefhyde — COMMENTED** [review]
LGTM, pending review from @Oats87

**Review by jakefhyde — DISMISSED** [review]

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:45]:
```suggestion
	rkeClusters provcontrollers.ClusterClient
```

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:9]:
```suggestion
	provcontrollers "github.com/rancher/rancher/pkg/generated/controllers/provisioning.cattle.io/v1"
```

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go
Comment (jakefhyde) [reply pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:45]: Do we need the client here or will the cache suffice?

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller_test.go
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/rke2/bootstrap/controller_test.go:15]:
```suggestion
	provcontrollers "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
```

**Review by jakefhyde — CHANGES_REQUESTED** [review]
Couple nits

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go
Comment (HarrisonWAffel) [reply pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:45]: I tried to use a cache but it was always missing the cluster object, so a client seemed like the only option. If you have any tips on why that might be I'm all ears, but using `clients.Provisioning.Cluster().Cache()` seemed to be updating fast enough for us to use it here.

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller_test.go
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/rke2/bootstrap/controller_test.go:15]:
```suggestion
	provv1 "github.com/rancher/rancher/pkg/apis/provisioning.cattle.io/v1"
```
My bad, should've clarified this.

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller_test.go
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/rke2/bootstrap/controller_test.go:8]:
```suggestion
	provcontrollers "github.com/rancher/rancher/pkg/generated/controllers/provisioning.cattle.io/v1"
```

This should probably stay `provcontrollers` so it's clear which import is for the controllers and which is for the types.

**Review by jakefhyde — DISMISSED** [review]

**Review by Oats87 — CHANGES_REQUESTED** [review]

File: pkg/provisioningv2/rke2/installer/installer.go
Comment (HarrisonWAffel) [reply pkg/provisioningv2/rke2/installer/installer.go]: I've updated it so that the hooks are injected right after the shebang, so we don't mess with the `exit 0` call at all.

**Comment by jakefhyde** [comment]
@Oats87 @HarrisonWAffel We could also add the hooks as a proper field on the bootstrap object: https://github.com/rancher/rancher/blob/77f87c498b1517dcca5ae7fcf707dc8ebe7a95db/pkg/apis/rke.cattle.io/v1/bootstrap.go#L17-L19
Here is an example of how upstream does this:
https://github.com/kubernetes-sigs/cluster-api/blob/c1730d3cb9344bff059d92cc9a18b6b2a6faab04/bootstrap/kubeadm/api/v1alpha4/kubeadmconfig_types.go#L63
https://github.com/kubernetes-sigs/cluster-api/blob/6a4e90cad62de2f4ec0f48ed92f94d348ed41035/bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go#L486

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go
Comment (jakefhyde) [inline pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:142]: I feel like directly accessing the provisioning cluster may not be the most appropriate way to do this. It's the simplest, but if we want to stick closer to upstream, we would refactor this to create a bootstrap template per pool: https://github.com/rancher/rancher/blob/release/v2.7/pkg/controllers/provisioningv2/rke2/provisioningcluster/template.go#L235. cc @Oats87

**Review by jakefhyde — CHANGES_REQUESTED** [review]

**Comment by Oats87** [comment]
> @Oats87 @HarrisonWAffel We could also add the hooks as a proper field on the bootstrap object:
> 
> https://github.com/rancher/rancher/blob/77f87c498b1517dcca5ae7fcf707dc8ebe7a95db/pkg/apis/rke.cattle.io/v1/bootstrap.go#L17-L19
> 
> Here is an example of how upstream does this:
> https://github.com/kubernetes-sigs/cluster-api/blob/c1730d3cb9344bff059d92cc9a18b6b2a6faab04/bootstrap/kubeadm/api/v1alpha4/kubeadmconfig_types.go#L63
> https://github.com/kubernetes-sigs/cluster-api/blob/6a4e90cad62de2f4ec0f48ed92f94d348ed41035/bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go#L486

Yeah, this is definitely something we could do. I dislike how it's specifically "preKubeadm" i.e. specifically kubeadm, but that doesn't necessarily mean the idea itself isn't bad. 


On a semi-related note, @HarrisonWAffel did we have design discussion where we formally decided on using "hooks"? I think "hook" might be too generic of a field for what these specifically are. they really are "preHook"s.

File: pkg/controllers/provisioningv2/rke2/bootstrap/controller.go
Comment (Oats87) [reply pkg/controllers/provisioningv2/rke2/bootstrap/controller.go:142]: Yes, the most ideal manner of fixing this is to create a bootstrap template per pool; however, if we make this change we need to figure out how to make the change in such a way where we do not cause reprovisioning of existing machine pools when we change the name of the bootstrap template, and we need to clean up the old bootstrap templates. If possible, I really would like to clean up that behavior.

I'm not sure if making that change makes the scope of adding these prehooks too big for the given milestone though, but I would much prefer this method. 

cc: @snasovich

**Comment by Oats87** [comment]
@HarrisonWAffel @jakefhyde @snasovich before we merge this PR, we should discuss this in an afterparty item (or more realistically, in a design discussion). @jakefhyde brings up some good points about the implementation of this in our codebase vs. how upstream "would" do it.

**Pull request closed without merging**
